Add Miraie AC integration

milothomas opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->


## Type of change
- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/28022

## Checklist
- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

milothomas committed:
WIP - Integration now shows all devices

milothomas committed:
Cleanup and fixes

milothomas committed:
Merge branch 'home-assistant:dev' into miraie-ac

milothomas committed:
dev checklist completed.

milothomas committed:
Merge branch 'home-assistant:dev' into miraie-ac

milothomas committed:
Config flow handles auth errors

milothomas committed:
Merge branch 'miraie-ac' of https://github.com/milothomas/home-assistant into miraie-ac

milothomas committed:
Merge branch 'dev' into miraie-ac

milothomas committed:
Merge branch 'dev' into miraie-ac

milothomas committed:
Merge branch 'dev' into miraie-ac

milothomas committed:
Merge branch 'dev' into miraie-ac

milothomas committed:
Added tests

milothomas committed:
Merge branch 'miraie-ac' of https://github.com/milothomas/home-assistant into miraie-ac

milothomas committed:
Merge branch 'dev' into miraie-ac

milothomas committed:
Merge branch 'dev' into miraie-ac

milothomas committed:
Merge branch 'dev' into miraie-ac

schelv reviewed (commented):
Hi!
Seems like a useful addition!
I've added some comments on things that might be improved.
Hopefully they are useful! =)

I see you use this pip package that you created: https://pypi.org/project/py-miraie-ac/
The linked github project gives a 404 error.
Maybe the project is still private?

milothomas commented on homeassistant/components/miraie_ac/config_flow.py:
```diff
@@ -0,0 +1,80 @@
+        if user_input is not None:
+            try:
+                info = await validate_input(self.hass, user_input)
+            except CannotConnect:
+                errors["base"] = "cannot_connect"
```
No, I have removed this section and associated test

milothomas commented on homeassistant/components/miraie_ac/const.py:
```diff
@@ -0,0 +1,5 @@
+"""Constants for the MirAIe AC integration."""
+
+DOMAIN = "miraie_ac"
+CONFIG_KEY_USER_ID = "mobile"
```
There are three different types "User ID" - mobile number, email and username - commonly referred to as User ID

milothomas replied:
There are 3 different ways to login, mobile number, email or a userid. The app currently only supports mobile number but we could add the other options in a future update.

edenhaus replied:
```suggestion
CONFIG_KEY_USER_ID = "mobile"
```

milothomas commented on homeassistant/components/miraie_ac/config_flow.py:
```diff
@@ -0,0 +1,80 @@
+                errors["base"] = "cannot_connect"
+            except InvalidAuth:
+                errors["base"] = "invalid_auth"
+            except Exception:  # pylint: disable=broad-except
+                _LOGGER.exception("Unexpected exception")
+                errors["base"] = "unknown"
```
Updated message to show "try again later". Will have to check the logs to see what went wrong. Any suggestions to clean this up?

schelv replied:
Maybe something like "try again later, or report the problem <somewhere> with <needed information>"?
Ideally this code is never encountered.

milothomas commented on homeassistant/components/miraie_ac/config_flow.py:
```diff
@@ -0,0 +1,80 @@
+
+STEP_USER_DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONFIG_KEY_USER_ID): str,
```
Yes, I have added a validation function and added tests.

milothomas commented on homeassistant/components/miraie_ac/config_flow.py:
```diff
@@ -0,0 +1,80 @@
+            raise InvalidAuth from ex
+
+        _LOGGER.debug("Connection successful!")
+        # Return info that you want to store in the config entry.
+        return {"title": "MirAIe"}
```
I believe this is how the file is generated after scaffolding. Do we need to change this?

schelv replied:
You mean here? or in the scaffold? 😅
change here is not really super necessary but it's a tiny thing; having the names of methods match with what they do/return is convenient.

milothomas commented on homeassistant/components/miraie_ac/config_flow.py:
```diff
@@ -0,0 +1,80 @@
+        try:
+            _LOGGER.debug("Initializing MirAIe API")
+            await api.initialize()
+        except AuthException as ex:
+            _LOGGER.error("Invalid user ID or password")
```
Yes, there is an http call to login and the exception is raised if it comes back with an error.

schelv replied:
But there are many other possible errors right?
The webservice could be totally unresponsive.
It might return a response with status code 500.

Will these also show up as "Invalid user ID or password" to the user?

milothomas commented on homeassistant/components/miraie_ac/__init__.py:
```diff
@@ -0,0 +1,44 @@
+    login_id = entry.data[CONFIG_KEY_USER_ID]
+    password = entry.data[CONFIG_KEY_PASSWORD]
+
+    api = MirAIeAPI(auth_type=AuthType.MOBILE, login_id=login_id, password=password)
```
This has not been tested yet, but it may be possible in future. Although the app only currently supports mobile number.

milothomas commented on homeassistant/components/miraie_ac/climate.py:
```diff
@@ -0,0 +1,372 @@
+FAN_MODE_MAP_TO_MIRAIE = {
+    FAN_AUTO: MirAIeFanMode.AUTO,
+    FAN_HIGH: MirAIeFanMode.HIGH,
+    FAN_LOW: MirAIeFanMode.LOW,
+    FAN_MEDIUM: MirAIeFanMode.MEDIUM,
+    FAN_OFF: MirAIeFanMode.QUIET,
+}
+
+
+SUPPORTED_HVAC_MODES = [
+    HVACMode.AUTO,
+    HVACMode.COOL,
+    HVACMode.OFF,
+    HVACMode.DRY,
+    HVACMode.FAN_ONLY,
+]
+
+SUPPORTED_PRESET_MODES = [PRESET_NONE, PRESET_ECO, PRESET_BOOST]
+SUPPORTED_FAN_MODES = [
+    FAN_AUTO,
+    FAN_LOW,
+    FAN_MEDIUM,
+    FAN_HIGH,
+    FAN_OFF,
+]
+
+SUPPORTED_SWING_MODES = [
+    SWING_OFF,
+    SWING_VERTICAL,
+    SWING_HORIZONTAL,
+    SWING_BOTH,
+]
+
+SUPPORTED_FEATURES = (
+    ClimateEntityFeature.TARGET_TEMPERATURE
+    | ClimateEntityFeature.FAN_MODE
+    | ClimateEntityFeature.PRESET_MODE
+    | ClimateEntityFeature.SWING_MODE
+)
+
```
Yea, moved all the constants to the const.py file.
Yes, these are all supported by the MirAIe AC devices.

milothomas commented on homeassistant/components/miraie_ac/climate.py:
```diff
@@ -0,0 +1,372 @@
+        self._attr_swing_modes = SUPPORTED_SWING_MODES
+        self._attr_supported_features = SUPPORTED_FEATURES
+        self._attr_fan_mode = FAN_OFF
+        self._attr_max_temp = 30.0
+        self._attr_min_temp = 16.0
```
These are values set on the app. I've moved these to the constants.

schelv replied:
Ok clear. maybe you can add a comment with where you got the numbers originally.

milothomas committed:
Review comments addressed

milothomas committed:
Merge branch 'miraie-ac' of https://github.com/milothomas/home-assistant into miraie-ac

milothomas committed:
Review comments addressed

milothomas commented:
> Hi! Seems like a useful addition! I've added some comments on things that might be improved. Hopefully they are useful! =)
> 
> I see you use this pip package that you created: https://pypi.org/project/py-miraie-ac/ The linked github project gives a 404 error. Maybe the project is still private?

Yes, it was still private. I've changed it now: https://github.com/milothomas/py-miraie-ac/

schelv commented on homeassistant/components/miraie_ac/config_flow.py:
```diff
@@ -44,6 +53,12 @@ async def validate_input(hass: HomeAssistant, data: dict[str, Any]) -> dict[str,
 
 
+def _validate_mobile_number(mobile_number):
+    pattern = r"^\+91\d{10}$"
```
+91? is that the country code of india?

milothomas replied:
Yes, I think these devices are only available in India at present

schelv reviewed (commented):
.

milothomas committed:
Review comments addressed

milothomas committed:
Merge branch 'dev' into miraie-ac

milothomas commented:
hey @schelv, thanks for the review. I have updated the library to raise a couple of different exceptions now and these are handled accordingly in the config flow and updated the tests. Let me know what you think.

milothomas committed:
Merge branch 'dev' into miraie-ac

schelv reviewed (approved):
I think it looks good!

milothomas committed:
Merge branch 'dev' into miraie-ac

milothomas committed:
Merge branch 'dev' into miraie-ac

milothomas committed:
Merge branch 'dev' into miraie-ac

milothomas committed:
Merge branch 'dev' into miraie-ac

milothomas committed:
Merge branch 'dev' into miraie-ac

milothomas committed:
Merge branch 'dev' into miraie-ac

milothomas committed:
Merge branch 'dev' into miraie-ac

milothomas committed:
Merge branch 'dev' into miraie-ac

milothomas committed:
Merge branch 'dev' into miraie-ac

milothomas committed:
Merge branch 'dev' into miraie-ac

milothomas committed:
Merge branch 'dev' into miraie-ac

milothomas committed:
Merge branch 'dev' into miraie-ac

milothomas committed:
Merge branch 'dev' into miraie-ac

edenhaus commented on homeassistant/components/miraie_ac/config_flow.py:
```diff
@@ -0,0 +1,107 @@
+                raise ValidationError()
+
+            await api.initialize()
+        except ValidationError:
+            _LOGGER.error("Invalid Mobile Number")
+            raise
+        except AuthException:
+            _LOGGER.error("Invalid user ID or password")
+            raise
+        except ConnectionException:
+            _LOGGER.error("Error connecting to the MirAIe services")
+            raise
+        except MobileNotRegisteredException:
+            _LOGGER.error("Mobile number not registered with MirAIe")
+            raise
```
As you reraise the exceptions, the logging can also be moved to the parent method, where you handle the exceptions.

edenhaus commented on homeassistant/components/miraie_ac/config_flow.py:
```diff
@@ -0,0 +1,107 @@
+        try:
+            _LOGGER.debug("Initializing MirAIe API")
+
+            if not _validate_mobile_number(login_id):
+                raise ValidationError()
```
Validating the mobile phone number should be done before creating the API object.

edenhaus commented on .coveragerc:
```diff
@@ -1563,7 +1563,8 @@ omit =
     homeassistant/components/electrasmart/climate.py
     homeassistant/components/electrasmart/__init__.py
-
+    homeassistant/components/miraie_ac/climate.py
```
Can you please insert your lines in the correct sorted position?

We know that it is not correctly sorted but don't make it worst :)

edenhaus commented on homeassistant/components/miraie_ac/config_flow.py:
```diff
@@ -0,0 +1,107 @@
+STEP_USER_DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONFIG_KEY_USER_ID): str,
+        vol.Required(CONFIG_KEY_PASSWORD): str,
```
```suggestion
        vol.Required(CONF_PASSWORD): str,
```

Please use the available constants from `homeassistant.const`.

edenhaus commented on homeassistant/components/miraie_ac/manifest.json:
```diff
@@ -0,0 +1,13 @@
+  "name": "MirAIe AC",
+  "codeowners": ["@milothomas"],
+  "config_flow": true,
+  "dependencies": [],
```
Please remove all empty keys.

edenhaus commented on homeassistant/components/miraie_ac/strings.json:
```diff
@@ -0,0 +1,22 @@
+    "step": {
+      "user": {
+        "data": {
+          "mobile": "Mobile Number (e.g. +91xxxxxxxxxx)",
```
```suggestion
          "mobile": "Mobile Number",
```

edenhaus commented on homeassistant/components/miraie_ac/config_flow.py:
```diff
@@ -0,0 +1,107 @@
+                _LOGGER.exception("Unexpected exception")
+                errors["base"] = "unknown"
+            else:
+                return self.async_create_entry(title=info["title"], data=user_input)
```
I assume only one config entry per mobile number is possible. Therefore please set the uniwue_id and abort if already a config entry exists with the given number.

edenhaus commented on homeassistant/components/miraie_ac/climate.py:
```diff
@@ -0,0 +1,303 @@
+
+        entities = list(map(MirAIeClimateEntity, devices))
+        async_add_entities(entities)
+    except Exception as exp:
```
Handling the bare exception is only allowed in the config flow. Please handle only exceptions raised by the API.

edenhaus commented on homeassistant/components/miraie_ac/climate.py:
```diff
@@ -0,0 +1,303 @@
+
+class MirAIeClimateEntity(ClimateEntity):
+    """Define MirAIe Climate."""
+
```
```suggestion

    _attr_hvac_modes = SUPPORTED_HVAC_MODES
    _attr_preset_modes = SUPPORTED_PRESET_MODES
    _attr_fan_modes = SUPPORTED_FAN_MODES
    _attr_swing_modes = SUPPORTED_SWING_MODES
    _attr_supported_features = SUPPORTED_FEATURES
    _attr_fan_mode = FAN_OFF
    _attr_max_temp = MAX_TEMP
    _attr_min_temp = MIN_TEMP
    _attr_temperature_unit = UnitOfTemperature.CELSIUS
    _attr_precision = PRECISION_WHOLE

```

edenhaus commented on homeassistant/components/miraie_ac/climate.py:
```diff
@@ -0,0 +1,303 @@
+
+    def __init__(self, device: MirAIeDevice) -> None:
+        """Initialize MirAIe climate entity."""
+        self._attr_hvac_modes = SUPPORTED_HVAC_MODES
+        self._attr_preset_modes = SUPPORTED_PRESET_MODES
+        self._attr_fan_modes = SUPPORTED_FAN_MODES
+        self._attr_swing_modes = SUPPORTED_SWING_MODES
+        self._attr_supported_features = SUPPORTED_FEATURES
+        self._attr_fan_mode = FAN_OFF
+        self._attr_max_temp = MAX_TEMP
+        self._attr_min_temp = MIN_TEMP
+        self._attr_temperature_unit = UnitOfTemperature.CELSIUS
+        self._attr_precision = PRECISION_WHOLE
```
```suggestion
```

milothomas replied:
hey @edenhaus, I don't understand. Do you want me to remove this? Could you elaborate please?

edenhaus commented on homeassistant/components/miraie_ac/climate.py:
```diff
@@ -0,0 +1,303 @@
+
+        _LOGGER.debug("MirAIe device added: %s", device.friendly_name)
+
+    @property
+    def should_poll(self) -> bool:
+        """Let Hass know that polling is not required."""
+        return False
```
Set it as class variable

edenhaus commented on homeassistant/components/miraie_ac/climate.py:
```diff
@@ -0,0 +1,303 @@
+        """Return the display name of this light."""
+        return self._friendly_name
+
+    @property
+    def icon(self) -> str | None:
+        """Return the icon to use on the frontend, if any."""
+        return "mdi:air-conditioner"
+
+    @property
+    def device_info(self) -> DeviceInfo:
+        """Return the device info."""
+        return DeviceInfo(
+            identifiers={(DOMAIN, self.device.device_id)},
+            name=self._friendly_name,
+            manufacturer=self.device.brand,
+            model=self.device.model_number,
+            sw_version=self.device.firmware_version,
+            suggested_area=self.device.area_name,
+        )
```
Set it as class variable

edenhaus commented on homeassistant/components/miraie_ac/climate.py:
```diff
@@ -0,0 +1,303 @@
+    def entity_state_changed_callback(self):
+        """Device status has changed, notify Hass that the entity state must be updated."""
+        _LOGGER.debug("%s - Hass State Updated", self._friendly_name)
+        self.hass.loop.call_soon_threadsafe(self.async_write_ha_state)
```
Some of the above properties can be set directly here instead.

milothomas replied:
Hi @edenhaus, could you elaborate? I am not sure what is the issue is. Thanks.

edenhaus replied:
Instead of creating all the properties above, you can directly set the variable.
Example temperature:
`self._attr_current_temperature = self.device.status.room_temp`

milothomas replied:
I want to have the debug logging so as to see when new updates are pushed. I've seen delays occasionally. 

I guess I could remove the logging, but Is there an advantage to doing it this way? - just trying to understand the reason behind this recommendation

edenhaus replied:
With `async_write_ha_state`, you inform the state machine about the new state. I don't see any advantage of logging the access of the properties too. How big are these delays?
Please note, that you call `async_write_ha_state` within `call_soon_threadsafe` and it's written in the docs:

> Schedule the callback [callback](https://docs.python.org/3/glossary.html#term-callback) to be called with args arguments at the next iteration of the event loop.

milothomas replied:
Ok, I've moved all the updates to an internal method.

edenhaus commented on homeassistant/components/miraie_ac/climate.py:
```diff
@@ -0,0 +1,303 @@
+        return False
+
+    @property
+    def name(self) -> str:
```
New integrations are required to implement the [new naming format](https://developers.home-assistant.io/docs/core/entity#has_entity_name-true-mandatory-for-new-integrations).

edenhaus reviewed (changes requested):
I have done a quick initial review.

Please fix the requested changes.

milothomas committed:
Merge branch 'home-assistant:dev' into miraie-ac

milothomas committed:
file paths moved to preserve alphabetical order

milothomas committed:
Merge branch 'home-assistant:dev' into miraie-ac

milothomas committed:
Merge remote-tracking branch 'refs/remotes/origin/miraie-ac' into miraie-ac

milothomas committed:
Use existing constants from homeassistant.const

milothomas committed:
Mobile number validated before creating API object

milothomas committed:
Clean up input validation and logging

milothomas committed:
Add country code to phone number if not present, updated tests.

milothomas committed:
Removed empty keys

milothomas committed:
Removed empty keys

milothomas committed:
Abort flow if mobile number is already configured

milothomas committed:
Bare exceptions are handled in config flow.

milothomas committed:
Use class variables where possible. New naming format applied.

milothomas committed:
Merge branch 'dev' into miraie-ac

milothomas committed:
Merge branch 'dev' into miraie-ac

milothomas committed:
Merge branch 'dev' into miraie-ac

milothomas committed:
Merge branch 'dev' into miraie-ac

milothomas committed:
Merge branch 'dev' into miraie-ac

milothomas committed:
Merge branch 'dev' into miraie-ac

edenhaus commented:
You don't need to merge the dev into your branch each day. For the future please use rebase instead :)

milothomas committed:
Properties are now set directly.

milothomas committed:
Merge branch 'dev' into miraie-ac

milothomas committed:
Merge branch 'home-assistant:dev' into miraie-ac

milothomas committed:
Moved climate-only constants to climate.py

milothomas committed:
Refactored to cleanup unnecessary catch and raise statements.

milothomas committed:
Removed redundant property

milothomas committed:
Updated dependency

milothomas committed:
Merge branch 'dev' into miraie-ac

milothomas committed:
Merge branch 'dev' into miraie-ac

milothomas commented:
> I have done a quick initial review.
> 
> Please fix the requested changes.

Thanks, all items have been fixed.

milothomas commented:
@edenhaus, just checking if there is anything pending from my side on this?

edenhaus reviewed (changes requested):
Please make sure the CI is passing

edenhaus committed:
Merge branch 'dev' into miraie-ac

edenhaus commented on homeassistant/components/miraie_ac/config_flow.py:
```diff
@@ -0,0 +1,111 @@
+        _LOGGER.debug("Connection successful!")
+
+        # Return info that you want to store in the config entry.
+        return {"title": f"MirAIe ({login_id})"}
```
```suggestion
        return {"title": login_id}
```

No need to specify again the integrations name

edenhaus commented on homeassistant/components/miraie_ac/config_flow.py:
```diff
@@ -0,0 +1,111 @@
+        """Handle the initial step."""
+        errors: dict[str, str] = {}
+        if user_input is not None:
+            phone_number = user_input[CONFIG_KEY_USER_ID]
+            formatted_phone_number = _add_country_code(phone_number)
+            user_input[CONFIG_KEY_USER_ID] = formatted_phone_number
+
+            await self.async_set_unique_id(formatted_phone_number)
```
```suggestion
            phone_number = _add_country_code(user_input[CONFIG_KEY_USER_ID])
            user_input[CONFIG_KEY_USER_ID] = phone_number

            await self.async_set_unique_id(phone_number)
```

edenhaus commented on homeassistant/components/miraie_ac/climate.py:
```diff
@@ -0,0 +1,304 @@
+) -> None:
+    """Add MirAIe AC devices."""
+    api: MirAIeAPI = hass.data[DOMAIN][entry.entry_id]
+    await api.initialize()
```
Why is the api not initialized in the init file?
Please move it there.

Could this function raise any errors?

edenhaus commented on homeassistant/components/miraie_ac/climate.py:
```diff
@@ -0,0 +1,304 @@
+    await api.initialize()
+    devices: list[MirAIeDevice] = api.devices
+
+    for device in devices:
+        _LOGGER.debug("Found MirAIe device: %s", device.friendly_name)
```
Why do we need this? If we add all devices anyway?
Adding is also logged in line 176

I suggest removing it
```suggestion
```

edenhaus commented on homeassistant/components/miraie_ac/climate.py:
```diff
@@ -0,0 +1,304 @@
+    def __init__(self, device: MirAIeDevice) -> None:
+        """Initialize MirAIe climate entity."""
+
+        self.device = device
```
```suggestion
        self._device = device
```

milothomas committed:
Merge branch 'dev' into miraie-ac

milothomas committed:
PR review fixes

milothomas committed:
Update tests

milothomas committed:
Merge branch 'dev' into miraie-ac

milothomas committed:
Bumped py-miraie-ac

milothomas committed:
Merge branch 'home-assistant:dev' into miraie-ac

edenhaus commented on homeassistant/components/miraie_ac/climate.py:
```diff
@@ -0,0 +1,299 @@
+        self._attr_swing_mode = self._map_swing_mode(v_swing_mode, h_swing_mode)
+
+        # hvac mode
+        power_mode = self._device.status.power_mode
+        hvac_mode = self._device.status.hvac_mode
+        self._attr_hvac_mode = (
+            HVACMode.OFF
+            if power_mode == MirAIePowerMode.OFF
+            else HVAC_MODE_MAP_TO_HASS[hvac_mode]
+        )
```
```suggestion
        if power_mode == MirAIePowerMode.OFF:
            self._attr_hvac_mode = HVACMode.OFF
        else:
            hvac_mode = self._device.status.hvac_mode
            self._attr_hvac_mode = HVAC_MODE_MAP_TO_HASS[hvac_mode]
```

edenhaus commented on homeassistant/components/miraie_ac/config_flow.py:
```diff
@@ -0,0 +1,110 @@
+)
+
+
+async def validate_input(hass: HomeAssistant, data: dict[str, Any]) -> dict[str, Any]:
```
```suggestion
async def validate_input(hass: HomeAssistant, data: dict[str, Any]) -> None:
```

edenhaus commented on homeassistant/components/miraie_ac/config_flow.py:
```diff
@@ -0,0 +1,110 @@
+            self._abort_if_unique_id_configured()
+
+            try:
+                info = await validate_input(self.hass, user_input)
```
```suggestion
                await validate_input(self.hass, user_input)
```

edenhaus commented on homeassistant/components/miraie_ac/config_flow.py:
```diff
@@ -0,0 +1,110 @@
+        _LOGGER.debug("Initializing MirAIe API")
+        await api.initialize()
+        _LOGGER.debug("Connection successful!")
+
+        # Return info that you want to store in the config entry.
+        return {"title": login_id}
```
```suggestion
```

edenhaus commented on homeassistant/components/miraie_ac/config_flow.py:
```diff
@@ -0,0 +1,110 @@
+                _LOGGER.exception("Unexpected exception")
+                errors["base"] = "unknown"
+            else:
+                return self.async_create_entry(title=info["title"], data=user_input)
```
```suggestion
                return self.async_create_entry(title=phone_number, data=user_input)
```

We have already the phone_number in a local variable

edenhaus commented on tests/components/miraie_ac/test_config_flow.py:
```diff
@@ -0,0 +1,220 @@
+        result3["flow_id"],
+        {CONFIG_KEY_USER_ID: MOBILE, CONF_PASSWORD: "P@ssw0rD"},
+    )
+    assert result4["type"] == data_entry_flow.FlowResultType.ABORT
```
Please check the error

edenhaus commented on tests/components/miraie_ac/test_config_flow.py:
```diff
@@ -0,0 +1,220 @@
+    assert result4["type"] == data_entry_flow.FlowResultType.ABORT
+
+
+async def test_form_invalid_auth(hass: HomeAssistant) -> None:
+    """Test we handle invalid auth."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+
+    with patch(
+        "homeassistant.components.miraie_ac.config_flow.MirAIeAPI.initialize",
+        side_effect=AuthException,
+    ):
+        result2 = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            {CONFIG_KEY_USER_ID: "+919876543219", CONF_PASSWORD: "P@ssw0rD"},
+        )
+
+    assert result2["type"] == FlowResultType.FORM
+    assert result2["errors"] == {"base": "invalid_auth"}
+
+
+async def test_form_cannot_connect(hass: HomeAssistant) -> None:
+    """Test we handle cannot connect error."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+
+    with patch(
+        "homeassistant.components.miraie_ac.config_flow.MirAIeAPI.initialize",
+        side_effect=ConnectionException,
+    ):
+        result2 = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            {CONFIG_KEY_USER_ID: "+919876543219", CONF_PASSWORD: "P@ssw0rD"},
+        )
+
+    assert result2["type"] == FlowResultType.FORM
+    assert result2["errors"] == {"base": "cannot_connect"}
```
Please make a parameterized test out of it

edenhaus commented on tests/components/miraie_ac/test_config_flow.py:
```diff
@@ -0,0 +1,220 @@
+    assert result2["type"] == FlowResultType.FORM
+    assert result2["errors"] == {"base": "mobile_not_registered"}
+
+
+async def test_form_invalid_mobile(hass: HomeAssistant) -> None:
+    """Test we handle cannot connect error."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+
+    with patch(
+        "homeassistant.components.miraie_ac.config_flow.MirAIeAPI.initialize",
+        side_effect=ValidationError,
+    ):
+        result2 = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            {CONFIG_KEY_USER_ID: "+9198765432", CONF_PASSWORD: "P@ssw0rD"},
+        )
+
+    assert result2["type"] == FlowResultType.FORM
+    assert result2["errors"] == {"base": "invalid_mobile"}
+
+
+async def test_form_unhandled_exception(hass: HomeAssistant) -> None:
+    """Test we handle cannot connect error."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+
+    with patch(
+        "homeassistant.components.miraie_ac.config_flow.MirAIeAPI.initialize",
+        side_effect=Exception,
+    ):
+        result2 = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            {CONFIG_KEY_USER_ID: "+919876543219", CONF_PASSWORD: "P@ssw0rD"},
+        )
+
+    assert result2["type"] == FlowResultType.FORM
+    assert result2["errors"] == {"base": "unknown"}
```
For these tests too

edenhaus commented on tests/components/miraie_ac/test_config_flow.py:
```diff
@@ -0,0 +1,220 @@
+    assert result2["type"] == FlowResultType.CREATE_ENTRY
+    assert result2["title"] == MOBILE
+    assert result2["data"] == {
+        CONFIG_KEY_USER_ID: MOBILE,
+        CONF_PASSWORD: "P@ssw0rD",
+    }
+    assert len(mock_setup_entry.mock_calls) == 1
+
+
+async def test_form_no_country_code(
+    hass: HomeAssistant, mock_setup_entry: AsyncMock
+) -> None:
+    """Test we get the form."""
+    MOBILE = "9876543219"
+
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN,
+        context={"source": config_entries.SOURCE_USER},
+    )
+
+    assert result["type"] == FlowResultType.FORM
+    assert result["errors"] == {}
+
+    with patch(
+        "homeassistant.components.miraie_ac.config_flow.MirAIeAPI.initialize",
+        return_value=None,
+    ):
+        result2 = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            {CONFIG_KEY_USER_ID: MOBILE, CONF_PASSWORD: "P@ssw0rD"},
+        )
+        await hass.async_block_till_done()
+
+    assert result2["type"] == FlowResultType.CREATE_ENTRY
+    assert result2["title"] == f"{COUNTRY_CODE}{MOBILE}"
+    assert result2["data"] == {
+        CONFIG_KEY_USER_ID: f"{COUNTRY_CODE}{MOBILE}",
+        CONF_PASSWORD: "P@ssw0rD",
+    }
+    assert len(mock_setup_entry.mock_calls) == 1
```
Please make another parameterized test for it

edenhaus commented on homeassistant/components/miraie_ac/__init__.py:
```diff
@@ -0,0 +1,46 @@
+    password = entry.data[CONF_PASSWORD]
+
+    api = MirAIeAPI(auth_type=AuthType.MOBILE, login_id=login_id, password=password)
+    await api.initialize()
```
The library uses `paho_mqtt`, which is synced code and should not run in the event loop. Please run the code in an executor.
`api.initialize()` calls internally `broker.connect`, blocking the event loop until a connection is made.
Please run the broker connect in an executor.

edenhaus commented on homeassistant/components/miraie_ac/manifest.json:
```diff
@@ -0,0 +1,9 @@
+  "config_flow": true,
+  "documentation": "https://www.home-assistant.io/integrations/miraie_ac",
+  "iot_class": "cloud_push",
+  "requirements": ["py-miraie-ac==0.1.10"]
```
On https://github.com/milothomas/py-miraie-ac/blob/main/setup.cfg the version is still `0.1.9` therefore I can't make the sanity check for the specified version. Please update your GitHub repo to include the specified version

milothomas committed:
Merge branch 'home-assistant:dev' into miraie-ac

milothomas committed:
Merge branch 'dev' into miraie-ac
